lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,277 @@
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, credit_card, options)
+        post[:card] = {}
+        if options[:token].present?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = options[:token][:value]
+        else
+          if credit_card.respond_to?(:number)
+            post[:card][:expirationDate] = "0#{credit_card.month}#{credit_card.year.to_s[2..4]}"
+            post[:card][:number] = credit_card.number
+            post[:card][:entryMode] = options[:entry_mode]
+            post[:card][:present] = options[:present]
+            post[:card][:securityCode] = {}
+            if options[:security_code].present?
+              post[:card][:securityCode][:indicator] = options[:security_code][:indicator]
+              post[:card][:securityCode][:value] = options[:security_code][:value]
+            end
+          end
+          end
+
+      end
+
+      def add_customer(post, options)
+        if address = options[:customer]
+          post[:customer] = {}
+          post[:customer][:addressLine1] = address[:address_line1] if address[:address_line1]
+          post[:customer][:firstName] = address[:first_name]
+          post[:customer][:lastName] = address[:last_name]
+          post[:customer][:postalCode] = address[:postal_code]
+        end
+      end
+
+      def commit(action, parameters, headers)
+        url_postfix = action == 'accesstoken' ? 'credentials' : 'transactions'
+        url = (test? ? "#{test_url}#{url_postfix}/#{action}" : "#{live_url}#{url_postfix}/#{action}")
+        response = parse(ssl_post(url, parameters.to_json, headers))
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response, action),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def parse(body)
+        return {} if body == ''
+
+        JSON.parse(body)
+      end
+
+      def message_from(response)
+        success_from(response) ? 'Transaction successful' : response['longText']
+      end
+
+      def error_code_from(response)
+        return unless success_from(response)
+
+        STANDARD_ERROR_CODE_MAPPING[response['primaryCode']]
+      end
+
+      def authorization_from(response, action)
```

Comment:
update this method so that it fetches authorization token from authorize/capture to pass them on capture,void/refund respectively.